### ISSUE TITLE
Polymorphism support? and an error in Unity

### ISSUE THREAD

**Issue by nuverian** [rpgmaker/NetJSON#97]
Hello,

Let me start by saying awesome work with NetJSON.
I have 2 issues/questions if you don't mind.

1. Is polymorphism somehow supported?

2. Also, when I try to serialize a Vector3 in unity, I get this error:
```
InvalidProgramException: Invalid IL code in Vector3Class:WriteVector3 (UnityEngine.Vector3,System.Text.StringBuilder): IL_0045: call      0x0a000004
```

Thanks!!

**Comment by rpgmaker** [comment]
It should be supported. As for the error with vector is looks like an error with struct. I will take a look at it and get back to you

Thanks,

**Comment by nuverian** [comment]
Thank you very much :)

**Comment by nuverian** [comment]
When you say that polymorphism should be supported do you mean that it is right now but something is going wrong, or that it will in the future?
Thanks.

**Comment by rpgmaker** [comment]
The serialization aspect is supported. But the deserialization is a different story since you will need to know the type. 

Thanks,

**Comment by nuverian** [comment]
Hey.
Hmm. Maybe i'm doing something wrong, but it doesn't seem to serialize the actual type of the object. Here is a quick test:
```
public class Vehicle{
	public string name;
}

public class Plane : Vehicle{
	public float maxAttitude;
}

public class NetJsonTest : MonoBehaviour {
	
	public Vehicle vehicle;
	public string json;
	
	void Start () {
		vehicle = new Plane{name = "Boing", maxAttitude = 1000f};
		json = NetJSON.NetJSON.Serialize<Vehicle>(vehicle);
		Debug.Log(json); //->> results in: {"name":"Boing"}

		vehicle = NetJSON.NetJSON.Deserialize<Vehicle>(json);
		Debug.Log(vehicle.GetType()); //->> results in: typeof(Vehicle)
	}
}
```

Also, does the RegisterTypeSerializer works? I tried creating one, but the serializer doesn't seem to be taking it into account.

Cheers and thanks!

**Comment by rpgmaker** [comment]
Register method is only taking into accout when using dictionary where the value type is object or when the data type been used is of object type.

You will need to use serialize that takes type to simulate what you want. Since the generic serialize uses the T type. You need to use Serialize(value, value.GetType()). You need to do same for deserialize. So you can create your own method that uses generic but internally calls GetType on the T obj.

Thanks,

**Comment by nuverian** [comment]
Thanks for the clarifications.
I'm still not sure if it's possible to serialize and deserialize a model like this, where nodes is a List<Node> and unknown types:
```
Graph{
	Nodes{
		NodeA{

		}
		NodeB{

		}
		NodeC{
			
		}
	}
}
```

Also, shouldn't there be a RegisterTypeDeserializer, to convert the string from the TypeSerializer back to a T object for custom converters?

Thanks!

**Comment by rpgmaker** [comment]
If the type is just IList without a generic constraint then it will not be able to figure it out since the values will be of type object since it kinda of like using ArrayList. As for having the converter to convert back to T type for each serialized type, I did not add that because I have a class that already try to infer the data type by using using regular expressions. And also it will be difficult to expose something such as a type deserializer since you really don't know the type would be without going through a series of test to determine what it should be just like I currently do with regular expressions. The best I can do is allow hooking to the logic so you can better predict the expected type based on the string what yourself since they are going to be primitive type. And any unknown type is always automatically converted to a dictionary string object.

Let me know if you have other ideas.

Thanks,

**Comment by rpgmaker** [comment]
Also. Is it possible to post the structure of vector3 as you currently have in your project for me to test it with what I have.

Thanks,

**Comment by nuverian** [comment]
Hey,

No, the list is a List<Node>. Here is the test I'm trying to successfully serialize/deserialize and fail to understand how to do so using NetJSON :) 
```
	public class NetJsonTest : MonoBehaviour {
		
		public Graph graph;
		public string json;
		
		void Start () {
			graph = new Graph{name = "my graph"};
			graph.nodes = new List<Node>();
			graph.nodes.Add(new NodeA{ number = 10f } );
			graph.nodes.Add(new NodeB{ text = "hello" });
			json = NetJSON.NetJSON.Serialize(typeof(Graph), graph);
		}
	}

	public class Graph{
		public string name;
		public List<Node> nodes;
	}

	public class Node{
		//public Vector2 pos;
		public float posx;
		public float posy;
	}

	public class NodeA : Node{
		public float number;
	}

	public class NodeB : Node{
		public string text;
	}
```
The result json is:
```
{"name":"my graph","nodes":[{},{}]}
```

Regarding custom type deserializers, some other json libraries, use a a meta-info added into the json when serializing, like: "$type:":"UnityEngine.Vector3". Thus if there was a custom deserializer, you would know which one to call for the user to deserialize. But maybe this is beyond the scope of NetJSON. I just drop an idea here.

Also, it would be great if the custom serializer was possible to be called not only for a Dictionary<string, object>, but whenever the type is encountered regardless.

The vector3 struct I use is the one of Unity library. (I'm not sure if I'm allowed to post the actual code since it's in a dll): http://docs.unity3d.com/ScriptReference/Vector3.html

There are some things that could be wrong with this Vector3 struct, but of course if there was the ability to create custom ser/deserializer, it could be solved like this :)
```
string Vector3Serializer(Vector3 vector){
	return string.Format("{0},{1},{2}", vector.x, vector.y, vector.z);
}

Vector3 Vector3Deserializer(string str){
	str = str.Split(',');
	return new Vector3(str[0], str[1], str[2]);
}
```

An extra idea would be an attribute to control what is seen by the serializer, like [NetJSON.Serializable] for types without custom serializer (and possibly in the future deserializer).

Just some ideas here :)

Thank you for your time and effort.
Cheers!!

**Comment by rpgmaker** [comment]
What version of netjson are you using? You might want to make sure that NetJSON.IncludeFields is set to true. If it is false then it will not serialize the values. The latest version should have the flag enables by default.

I will look into extending the register method for customizing how types are handled.

My main concern is to avoid adding functionality that will degrade performance.

Thanks,

**Comment by rpgmaker** [comment]
I will play with it tomorrow and see how much of polymorphism I can support without degrading the overall performance.

Thanks,

**Comment by nuverian** [comment]
Hey,

Yes, I understand that the goal of NetJSON is to be as fast as possible, but NetJSON is by far the faster serializer I've seen yet. So, I just think that adding a bit of convenience features like polymorphism, custom ser/deserializers, customization with attributes etc, even if at a small sacrifice of speed, it will be awesome, easier to use and will still be faster by far than any other serializer.

Then if these features degrade the performance a bit, you could add options to turn them on/off if people dont require them.

Just my thoughts on this :+1: 
Thanks!

**Comment by rpgmaker** [comment]
That is the plan. It will most likely be an option that will need to enabled in other to take advantage of it.

Thanks,

**Comment by nuverian** [comment]
Sounds great! Looking forward to what you make of it.
Thanks a lot for your time and effort :)

**Comment by rpgmaker** [comment]
I have fixed the initial issue with using vector3. It was caused by having indexer property. I am going to work on the poly issue and give you my progress as make strides.

Can you use the latest commit against unity and let me know if serialization and deserialization of vector3 works now?

Thanks,

**Comment by rpgmaker** [comment]
Just added NetJSON.IncludeTypeInformation which allows you to be able to serialize your Node object properly in the current commits ( https://github.com/rpgmaker/NetJSON/commit/6269ed1ed7838b5a23e153d8a67eb3d231464f14)

So you code looks more like this

 NetJSON.IncludeTypeInformation = true;

            var graph = new Graph { name = "my graph" };
            graph.nodes = new List<Node>();
            graph.nodes.Add(new NodeA { number = 10f });
            graph.nodes.Add(new NodeB { text = "hello" });
            var json = NetJSON.Serialize(graph);
            var jgraph = NetJSON.Deserialize<Graph>(json);


 public class Graph {
        public string name;
        public List<Node> nodes;
    }

    //Including the NetJSONKnownType is optionally but it is advise for better performance, without it
   //NetJSON will try to discover the Node subclasses itself
    [NetJSONKnownType(typeof(NodeA)), NetJSONKnownType(typeof(NodeB))]
    public class Node {
        //public Vector2 pos;
        public float posx;
        public float posy;
    }

    public class NodeA : Node {
        public float number;
    }

    public class NodeB : Node {
        public string text;
    }

Trying to add the logic to process the deserialization based on the $type when it is provided.

Can you test to let me know if the current code works so far?

Thanks,

**Comment by nuverian** [comment]
Hey,

Thats great! :)
I did some testing and the serialization works fine, yes. Good job!

Here are the only 2 cases I found out that it doesn't.

The first is relevant to generics:
```
public class NetJsonTest : MonoBehaviour {
	
	public Graph graph;
	public string json;
	
	void Start () {

		graph = new Graph{name = "my graph"};
		graph.nodes = new List<Node>();
		graph.nodes.Add(new Node<float>{ posx = 100f, value = 10f });
		graph.nodes.Add(new Node<string>{ value = "hello" });

		NetJSON.NetJSON.IncludeTypeInformation = true;

		json = NetJSON.NetJSON.Serialize(graph);
		//graph = NetJSON.NetJSON.Deserialize<Graph>(json);
	}
}

public class Graph{
	public string name;
	public List<Node> nodes;
}

public class Node{
	//public Vector2 pos;
	public float posx;
	public float posy;
}

public class Node<T> : Node{
	public T value;
}
```
If I add `[NetJSON.NetJSONKnownType(typeof(Node<float>)), NetJSON.NetJSONKnownType(typeof(Node<string>))]` above Node in the above case it works fine though.

------------
The other case is when having non-primitive types in an object data type:
```
public class NetJsonTest : MonoBehaviour {
	
	public List<object> objects = new List<object>();
	public string json;
	
	void Start () {

		objects.Add( 13f );
		objects.Add( new Node() );
		
		NetJSON.NetJSON.IncludeTypeInformation = true;
		json = NetJSON.NetJSON.Serialize(objects);
		//graph = NetJSON.NetJSON.Deserialize<Graph>(json);
	}
}
```
Which results in this: `[13,]`
Of course this is a bit less important I suppose.

Other than these 2 cases, the serialization seems to work great!
Cheers!

**Comment by rpgmaker** [comment]
The reason the object did not work is because the object type does not support complex objects. It only support primitive with the option to register your own. You can fake it by adding a register type method for node and call the serialize method from it.

Thanks,

**Comment by nuverian** [comment]
Ah, thanks for the clarification. That's fine!
I just mentioned it because I found it out. It's really not an actual issue :)

**Comment by rpgmaker** [comment]
No problem. You could see if it will actually work if you use the register type serializer.

Thanks,

**Comment by nuverian** [comment]
Yes, it works fine if I register a custom type serializer :)

**Comment by rpgmaker** [comment]
I am going to try to add the logic for deserializing the json after it contains the $type property. Doing so will most likely allow me to be able to deserialize data created from json.net and other libraries.

Once that is done, I will close this issue except you think everything you need has been satisfied.

Thanks,

**Comment by nuverian** [comment]
Hey,

That's awesome to hear! Looking forward to this.
Did you by any chance looked after the first case problem posted above with the generics?
https://github.com/rpgmaker/NetJSON/issues/97#issuecomment-164788568

The only other thing that comes in mind is the CustomTypeDeserializer so that we can take control over some types, but I could open a new issue if you like for that.

Thank you very much!!

**Comment by rpgmaker** [comment]
Can you post the detail again. Could not decipher which part of the generic
you are referring to.

Thanks,

**Comment by nuverian** [comment]
Yes. The problem is relevant to the new type info, when serializing a generic class that derives from a non-generic class like the following:
```
	abstract public class Node{

	}

	public class Node<T> : Node{
		public T value;
	}

	void Start () {
		var nodes = new List<Node>();
		nodes.Add( new Node<string>{ value = "hello"});
		nodes.Add( new Node<float>{ value = 13f });
		NetJSON.NetJSON.IncludeTypeInformation = true;
		var json = NetJSON.NetJSON.Serialize(nodes);
		//..
	}
```
In the above example, I get:
```
ArgumentNullException: Argument cannot be null.
Parameter name: obj
```

If I explicitely add the new attribute `[NetJSON.NetJSONKnownType(typeof(Node<float>))]`, the Node<float> does serialize correctly, but not without it.

Thanks.

**Comment by rpgmaker** [comment]
I see now. That is a little bit different because it will be difficult to infer the possible usage of the class generic constraint. I generate the code to create each type of mapping that are considered base class. I will have to play with that a little to figure with how best to solve it. In the meantime, i will try to focus on getting a solution for the deserialization of $type.

Thanks,

**Comment by rpgmaker** [comment]
What of the other reason it does not work with the generic is because i do not call Type.GetType on the object instance at runtime because of the cost of reflection.

Thanks,

**Comment by nuverian** [comment]
Thanks for the explanation.
Yeah, I suppose the only way would be to get the type of the instance if there is an instance. Since this needs to be done only in serialization and not in deserialization, I might add this in a fork after you find a solution for deserializing $type, which is certainly more important :)
Thanks! :+1:
